src/iterators/not_iterator.c

```diff
@@ -0,0 +1,237 @@
+  if (base->atEOF || base->lastDocId >= ni->maxDocId) {
+    base->atEOF = true;
+    return ITERATOR_EOF;
+  }
+
+  IteratorStatus rc;
+
+  if (ni->child->lastDocId == 0) {
+    // Read at least the first entry of the child
+    rc = ni->child->Read(ni->child);
+    if (rc == ITERATOR_TIMEOUT) return ITERATOR_TIMEOUT;
+  }
+  //Advance and see if it si valid
+  base->lastDocId++;
+
+  while (base->lastDocId <= ni->maxDocId) {
+    // I have the risk of missing one lastDocID if childEOF is not accurate (if the child requires another Read to set the flag)
+    if (base->lastDocId < ni->child->lastDocId || ni->child->atEOF) {
+      ni->timeoutCtx.counter = 0;
+      base->current->docId = base->lastDocId;
+      return ITERATOR_OK;
+    }
+    rc = ni->child->Read(ni->child);
+    if (rc == ITERATOR_TIMEOUT) return rc;
+    // Check for timeout with low granularity (MOD-5512)
+    //TODO(Joan): Ticket about magic number? Should it be a config?
+    if (TimedOut_WithCtx_Gran(&ni->timeoutCtx, 5000)) {
+      base->atEOF = true;
+      return ITERATOR_TIMEOUT;
+    }
+    // The lastDocID that we proposed is not valid, so we need to try another one
+    base->lastDocId++;
+  }
+  base->atEOF = true;
+  return ITERATOR_EOF;
+}
+
+/* Read from a NOT iterator - Optimized version, utilizing the `existing docs`
+ * inverted index. This is applicable only if the only or leftmost node of a
+ * query is a NOT node. We simply read until max docId, skipping docIds that
+ * exist in the child */
+static IteratorStatus NI_ReadSorted_O(QueryIterator *base) {
+  NotIterator *ni = (NotIterator *)base;
+  if (base->atEOF || base->lastDocId >= ni->maxDocId) {
+    base->atEOF = true;
+    return ITERATOR_EOF;
+  }
+
+  IteratorStatus rc;
+
+  if (ni->child->lastDocId == 0) {
+    // Read at least the first entry of the child
+    rc = ni->child->Read(ni->child);
+    if (rc == ITERATOR_TIMEOUT) return ITERATOR_TIMEOUT;
+  }
+  //Advance and see if it si valid
+  rc = ni->wcii->Read(ni->wcii);
+  if (rc == ITERATOR_TIMEOUT) {
+    base->atEOF = true;
+    return ITERATOR_TIMEOUT;
+  }
+  base->lastDocId = base->current->docId = ni->wcii->current->docId;
+
+  while (base->lastDocId <= ni->maxDocId) {
+    // I have the risk of missing one lastDocID if childEOF is not accurate (if the child requires another Read to set the flag)
+    if (base->lastDocId < ni->child->lastDocId || ni->child->atEOF) {
+      ni->timeoutCtx.counter = 0;
+      base->current->docId = base->lastDocId;
+      return ITERATOR_OK;
+    }
+    rc = ni->child->Read(ni->child);
+    if (rc == ITERATOR_TIMEOUT) return rc;
+    // Check for timeout with low granularity (MOD-5512)
+    //TODO(Joan): Ticket about magic number? Should it be a config?
+    if (TimedOut_WithCtx_Gran(&ni->timeoutCtx, 5000)) {
+      base->atEOF = true;
+      return ITERATOR_TIMEOUT;
+    }
+    // The lastDocID that we proposed is not valid, so we need to try another one
+    rc = ni->wcii->Read(ni->wcii);
```

Comment:
Unlike the non-optimized implementation, the wildcard iterator can surpass the child iterator here. We have to address that, otherwise we may miss relevant results